File: hamming-error-correction/src/main/java/com/pivovarit/hamming/vavr/HammingHelper.java
package com.pivovarit.hamming.vavr;

import io.vavr.collection.Stream;

class HammingHelper {

    int codewordSize(int msgLength) {
        return Stream.from(2, 1)
          .filter(r -> msgLength + r + 1 <= 1 << r)
          .map(r -> r + msgLength)
          .head();
    }

    Stream<Integer> getHammingCodewordIndices(int msgLength) {
        return Stream.from(0, 1)
          .take(codewordSize(msgLength));
    }

    String getParityBit(int codeWordIndex, BinaryString msg) {
        return parityIndicesSequence(codeWordIndex, codewordSize(msg.getValue().length()))
          .map(it -> getDataBit(it, msg))
          .map(Integer::valueOf)
          .reduce((a, b) -> (a + b) % 2)
          .toString();
    }

    String getDataBit(int ind, BinaryString input) {
        return Character.toString(
          input.getValue().charAt(ind - Integer.toBinaryString(ind).length()));
    }

    Stream<Integer> parityIndicesSequence(int startIndex, int endExclusive) {
        return Stream.from(startIndex, 1)
          .take(endExclusive - startIndex)
          .zipWithIndex()
          .filter(t -> (t._2 % ((2 * (startIndex + 1)))) < startIndex + 1)
          .map(t -> t._1)
          .drop(1);
    }
}

File: hamming-error-correction/src/main/kotlin/com/pivovarit/hamming/domain/encode/stateless/HammingHelper.kt
package com.pivovarit.hamming.domain.encode.stateless

import com.pivovarit.hamming.domain.BinaryString

internal class HammingHelper {

    fun codewordSize(msgLength: Int) = generateSequence(2, Int::inc)
      .first { r -> msgLength + r + 1 <= (1 shl r) } + msgLength

    fun getHammingCodewordIndices(msgLength: Int) = generateSequence(0, Int::inc)
      .take(codewordSize(msgLength))

    fun getParityBit(codeWordIndex: Int, msg: BinaryString) =
      parityIndicesSequence(codeWordIndex, codewordSize(msg.length))
        .map { getDataBit(it, msg).toInt() }
        .reduce { a, b -> a xor b }
        .toString()

    fun getDataBit(ind: Int, input: BinaryString) =
      input[ind - Integer.toBinaryString(ind).length].toString()

    fun parityIndicesSequence(startIndex: Int, endExclusive: Int) = generateSequence(startIndex, Int::inc)
      .take(endExclusive - startIndex)
      .filterIndexed { i, _ -> i % ((2 * (startIndex + 1))) < startIndex + 1 }
      .drop(1)
}